lib/telemetry_sampler/vm.ex

```diff
@@ -0,0 +1,72 @@
+defmodule Telemetry.Sampler.VM do
+  @moduledoc """
+  Collection of functions dispatching `Telemetry` events with Erlang VM metrics.
+
+  See documentation for `Telemetry.Sampler` to learn how to use these functions.
+  """
+
+  @doc """
+  Dispatches events with amount of memory dynamically allocated by the VM.
+
+  A single event is dispatched for each type of memory measured. Event name is always
+  `[:vm, :memory]`. Event metadata includes only a single key, `:type`, which corresponds to the
+  type of memory measured. Event value is the amount of memory of type given in metadata allocated
+  by the VM, in bytes.
+
+  The set of memory types may vary: see documentation for `:erlang.memory/0` to learn about possible
+  values.
+  """
+  @spec memory() :: :ok
+  def memory() do
+    :erlang.memory()
+    |> Enum.each(fn {type, size} ->
+      Telemetry.execute([:vm, :memory], size, %{type: type})
+    end)
+  end
+
+  @doc """
+  Dispatches an event with message queue length of the given process.
+
+  `process` can be a pid, atom, `{:global, term}` tuple or `{:via, module, term}` tuple. The event
+  is not dispatched if the `process`'s pid can't be resolved or the pid is not alive.
+
+  Event name is `[:vm, :message_queue_length]`. Event value is the measured message queue length.
+  Event metadata includes two keys: `:process`, which maps to `process` argument given to this
+  function, and `:pid`, which maps to the pid of the `process`.
+  """
+  @spec message_queue_length(pid() | atom() | {:global, term()} | {:via, module(), term()}) :: :ok
+  def message_queue_length(process) do
```

Comment:
Now that I see the examples, I am wondering if it would be better to have this used as the tuple in the general metrics specification. i.e. instead of `[:vm, :message_queue_len]`, we would tell people to do:

    {[:my_app, :repo, :message_queue_len], {Telemetry.Sampler, :message_queue_length, [MyApp]}}

This gives more flexibility on the name of the event, which I think is desired.

Reply:
Then maybe we could have two types of functions in `VM` module? One which would return the measured value (e.g. `message_queue_length`) and the other which would dispatch an event with measured value (e.g. `dispatch_message_queue_length`). What do you think?

This introduces additional maintenance burden, but at the same time we provide something which works out of the box (`dispatch..`) as well as customizable alternative.

Reply:
>  One which would return the measured value (e.g. message_queue_length)

I just realized that returning the measured value can be tricky because the process may not exist at this point.

I thought more about this and I realize that the reason we can't agree on the best API is because we are trying to design the feature before we have the use cases. So here are my suggestions:

1. Let's just have the MFA based samplers for now (and not the tuple ones)

2. Let's remove the message_queue_len thingy for now too. In the module docs, you can provide a simple example by hand

3. Let's hide the Telemetry.Sampler.VM module. We will still provide the `vm_measurements` helper, but we can document all keys directly in the `vm_measurements` function documentation and keep the module as an implementation detail

I think this is the minimum API we can provide and it will allow us to get unblocked and move forward. We can always add more conveniences later, for now it is probably best to keep it simple.

Thoughts?

Reply:
> we are trying to design the feature before we have the use cases

Agreed, I'm glad that you noticed that because we'd spend so much more time on this unnecessarily 😅 

I agree with all your points, let's do this and move forward :)